### DISCUSSION TITLE
Automatic PDF Exporting when Saving

### DISCUSSION THREAD

**Discussion by olazzzzz** [freeplane/freeplane#958] [Ideas]
Hello all,

As I understand, the topic of creating a mobile version of Freeplane is a long and complex one.
In my personal experience, whenever I want to consult a mindmap on my phone, I mostly use PDF versions (not editing, just viewing).
For this purpose, I think a nice option would be to have Freeplane exporting automatically a PDF when you save the mindmap, maybe to a common directory. Would this be possible to achieve, perhaps with an add-on or script (of which I have no idea about)?
Thank you!

**Comment by ldmpub** [comment]
Nice idea 👍, I also use the same workaround: generate a PDF file of the MindMap to read it on my Smartphone.

Compared to [DroidPlane](https://play.google.com/store/apps/details?id=ch.benediktkoeppel.code.droidplane&hl=fr) the PDF file allows me to perform text searches.

Your proposal is nice and the key features could be:
* (as you said) **transparent "double" save** (both the native *.mm and the *.PDF version of the MindMap)
* (personal addition) **custom folder for the .PDF file** ; because I personaly use an auto-sync tool to send the file quickly to my Smartphone and this Drive folder (shared among several devices) is different from my working folder (+ in this use case, particularly for iOS users, it's useless to have the *.mm file in the same synchronized directory because the *.mm file can not be opened on the mobile device).

**Reply by ldmpub** [reply]
An other use case: auto PDF export could be useful when you want to share a Mindmap with other people who do not want to install Freeplane to "view" your maps.

**Comment by euu2021** [comment]
Interesting! I would try a script like this:

Save the map:
`node.map.save(true)
`

And then use the export method, like in [this example](https://docs.freeplane.org/attic/old-mediawiki-content/Batch_Jobs.html#example-exporting-to-file). It seems that it's possible to choose specific branches to export, which is useful when dealing with large mindmaps.

To make it happen automatically in the background (like an autosave), I would try the ideas suggested at #544 .

**Reply by olazzzzz** [reply]
Thank you both for your replies and great suggestions!
I have pasted the instructions into the script and it seems to be working well! Although the path directory has to be written with inverted slashes, which I found weird(?).
Anyhow, one last feature I would like to implement, would be for all nodes to be unfolded before exporting to PDF. Is there a script line I could include in order to achieve this?

**Reply by bepolymathe** [reply]
Oh... great discussion. Can you share you compiled script @olazzzzz if it works with specific branches to export to pdf ? 
Export in several formats with pandoc (and pandoc option in attribute) would be a great addon 🤔

**Reply by euu2021** [reply]
>  one last feature I would like to implement, would be for all nodes to be unfolded before exporting to PDF

If you don't care if the nodes stay unfolded after the export, then insert this, before the exporting command:
`c.findAll().each{it.setFolded(false)}`

So, it would be like this:

```
node.map.save(true)
c.findAll().each{it.setFolded(false)}
boolean overwriteExistingFile = true
def filename = 'C:\\Users\\kauen\\Documents\\' + node.map.file.name.replaceFirst('.mm$', '.pdf')
// English localization
c.export(node.map, new File(filename), 'Portable Document Format (PDF) (.pdf)', overwriteExistingFile)
```

Now, if you want to keep the same folded state as before, I have 2 ideas:

- before exporting, create a Hash Map to save the fold state of each node. Unfold All. Export. Restore the state os each node
- Or, create a new map view, unfold all, export, close the map view.

I don't know which way would perform better.

---

> if it works with specific branches to export to pdf ?

When I saw the parameter `node.map` I thought it would be possible to point to specific nodes, but actually this parameter takes a map object, and not a node object. Here is the code, for the sake of curiosity:
![image](https://user-images.githubusercontent.com/77707706/214388070-ac7c124b-0910-4aa9-ba36-003c9bcb8246.png)

To export specific branches, I would try picking a characteristic of those branches (eg, style, icon, attribute) and do:

- find all nodes with characteristic X
- for each:
  - jump in
  - export
  - jump out

---
Later this week, I will try those ideas and report back.

**Reply by euu2021** [reply]
This actually worked better than I expected. So, to export the whole map unfolded:
```
node.map.save(true)
menuUtils.executeMenuItems([ 'NewMapViewAction' ])
c.findAll().each{it.setFolded(false)}
boolean overwriteExistingFile = true
def filename = 'C:\\Users\\kauen\\Documents\\' + node.map.file.name.replaceFirst('.mm$', '.pdf')
// English localization
c.export(node.map, new File(filename), 'Portable Document Format (PDF) (.pdf)', overwriteExistingFile)
menuUtils.executeMenuItems([ 'CloseAction' ])
```
---
Now, to export only a branch, all unfolded (let's say I want to export the branch of node ID_1861667403:

```
node.map.save(true)
menuUtils.executeMenuItems([ 'NewMapViewAction' ])
c.select(ID_1861667403)
menuUtils.executeMenuItems([ 'JumpInAction' ])
c.findAll().each{it.setFolded(false)}
boolean overwriteExistingFile = true
def filename = 'C:\\Users\\kauen\\Documents\\' + node.map.file.name.replaceFirst('.mm$', '.pdf')
// English localization
c.export(node.map, new File(filename), 'Portable Document Format (PDF) (.pdf)', overwriteExistingFile)
menuUtils.executeMenuItems([ 'CloseAction' ])
```

Just change `ID_1861667403` for your target node ID. 

To export multiple branches to multiple files, I think that simply repeating the code above (changing the exporting path) inside the script will work.

**Reply by olazzzzz** [reply]
Fantastic solution, precisely what I wanted! Thank you very much euu2021!
I just added a final 'node.map.save(true)' to make sure that the map gets saved in the end.
Thank you all, the program is great but the community is even better!

**Comment by olazzzzz** [comment]
While the script is working fine in most mindmaps, in the bigger/heavier ones, no PDF is actually exported and in the error log file I can read "java.lang.OutOfMemoryError: Java heap space". Can I ask what this means and if there is a way I can configure things to solve it?
In some other instances I was getting this error "Width (-1) and height (-1) cannot be <= 0", but since I relinked some missing images, it seems to have solved it.

**Reply by euu2021** [reply]
Try [this](https://docs.freeplane.org/attic/old-mediawiki-content/Troubleshooting.html?highlight=memory#freeplane-throws-out-of-memory-exception). 

I've seen this problem reported here and in the old forum, so maybe there are other solutions.
